common/hashing.h

```diff
@@ -0,0 +1,843 @@
+//
+// The algorithm used is most heavily based on [Abseil's hashing algorithm][1],
+// with some additional ideas and inspiration from the fallback hashing
+// algorithm in [Rust's AHash][2] and the FxHash function. However, there are
+// also *significant* changes introduced here.
+//
+// [1]: https://github.com/abseil/abseil-cpp/tree/master/absl/hash/internal
+// [2]: https://github.com/tkaitchuck/aHash/wiki/AHash-fallback-algorithm
+//
+// This hash algorithm does *not* defend against hash flooding. While it can be
+// viewed as "keyed" on the seed, it is expected to be possible to craft inputs
+// for some data types that cancel out the seed used and manufacture endlessly
+// colliding sets of keys. In general, this function works to be *fast* for hash
+// tables. If you need to defend against hash flooding, either directly use a
+// data structure with strong worst-case guarantees, or a hash table which
+// detects catastrophic collisions and falls back to such a data structure.
+//
+// This hash function is heavily optimized for *latency* over *quality*. Modern
+// hash tables designs can efficiently handle reasonable collision rates,
+// including using extra bits from the hash to avoid all efficiency coming from
+// the same low bits. Because of this, low-latency is significantly more
+// important for performance than high-quality, and this is heavily leveraged.
+// The result is that the hash codes produced *do* have significant avalanche
+// problems for small keys. The upside is that the latency for hashing integers,
+// pointers, and small byte strings (up to 32-bytes) is exceptionally low, and
+// essentially a small constant time instruction sequence.
+//
+// No exotic instruction set extensions are required, and the state used is
+// small. It does rely on being able to get the low- and high-64-bit results of
+// a 64-bit multiply efficiently.
+//
+// The function supports many typical data types such as primitives, string-ish
+// views, and types composing primitives transparently like pairs, tuples, and
+// array-ish views. It is also extensible to support user-defined types.
+//
+// The builtin support for string-like types include:
+// - `std::string_view`
+// - `std::string`
+// - `llvm::StringRef`
+// - `llvm::SmallString`
+//
+// This function supports heterogeneous lookup between all of the string-like
+// types. However, this is the only heterogeneous lookup support including for
+// the builtin in, standard, and LLVM types. Notably, each different size and
+// signedness integer type may hash differently for efficiency reasons. Hash
+// tables should pick a single integer type in which to manage keys and do
+// lookups.
+//
+// To add support for your type, you need to implement a customization point --
+// a free function that can be found by ADL for your type -- called `CarbonHash`
+// with the following signature:
+//
+// ```cpp
+// auto CarbonHash(const YourType& value, uint64_t seed) -> HashCode;
```

Comment:
I'd suggest calling this `CarbonHashValue`:
- It makes it a little clearer that this function has exactly the same contract as `HashValue`, it just doesn't support dispatching to arbitrary types.
- It helps discourage people from calling this function directly, because `HashValue` will be more concise.